atomix/cluster/src/main/java/io/atomix/cluster/messaging/impl/DefaultClusterEventService.java

```diff
@@ -403,7 +403,7 @@ InternalSubscriber localSubscriber() {
      *
      * @param subscription the subscription to unregister
      */
-    private synchronized CompletableFuture<Void> removeLocalSubscription(
+    private CompletableFuture<Void> removeLocalSubscription(
```

Comment:
On the local things I'm a bit unsure to remove that. This means it can happen that we remove one and concurrently add one; could this be a problem? That I end in a state where I actually subscribed but I'm not registered correctly?

But actually adding only happens on starting and removing on closing phase right? So probably this will never happen concurrently? :shrug:

Reply:
You are right. They have to be synchronized.

Reply:
I am also wondering if it would be better to remove all concurrent data structures and execute all public methods on an `eventServiceExecutor`. `broadcast` might be called frequently and it is now accessing two concurrent data structures - topics and remote subscribers. What do you think? I can do it in a different PR.

Reply:
Yes, I thought about that as well yesterday. That we normally do that with our actors, but since we have here also a single thread we probably could easily rebuild this kind of stuff. Or maybe we switch to our actor scheduler? Yes, would be fine if we do it on a follow up, just re-add the synchronize on the two methods for now.